clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -416,7 +428,13 @@ boolean joinGroupIfNeeded(final Timer timer) {
             }
 
             final RequestFuture<ByteBuffer> future = initiateJoinGroup();
-            client.poll(future, timer);
+
+            // if the flat is not set to true; we only try once and not block on the join result
+            if (waitUntilComplete)
+                client.poll(future, timer);
+            else
+                client.poll(timer);
```

Comment:
Do we want this to be `poll(0)`? Otherwise we're still blocking here.

Reply:
My rationale after thinking about this, is that it may be better to wait on at least some IOs to be done, especially if there are on-going fetch requests as well: we may complete writing a fetch request, or join / sync group request, or reading a response before returning.